BAR selection widget

moreginger opened the pull request:
Credit: this is largely a copy of https://github.com/beyond-all-reason/Beyond-All-Reason/blob/master/luaui/Widgets/gui_selectedunits_gl4.lua and https://github.com/beyond-all-reason/Beyond-All-Reason/blob/master/luaui/Widgets/gui_allySelectedUnits.lua

This screenshot shows how it looks on default settings. Green is local selection, which is also animated with a "breathe" (see video below). Blue is an ally selection - there's an option to disable this.

![image](https://github.com/moreginger/Zero-K/assets/1387874/d99f06ed-37ad-4273-aec4-95b943e23d32)

Ally/other (in replays) selections are shown in team color. Local selections are superimposed.

![image](https://github.com/moreginger/Zero-K/assets/1387874/2d4c1d4b-b05a-4e1f-b42b-5ef35bb21edd)

I included a few options but tried to keep things basic.

![image](https://github.com/moreginger/Zero-K/assets/1387874/a72888b6-5d96-4dc9-861f-55e3ef00917f)

This is intended to replace unit_shapes.lua. Can it do this, being GL4, or do we need both? The principle advantage is that it is considerably faster (~1% vs ~10% in this screenshot):

![image](https://github.com/moreginger/Zero-K/assets/1387874/37cf97a2-898e-4283-a003-4c0cfa3acc0f)

It differs from the BAR widgets in the following ways:

- The two widgets have been combined into one
- Highlights are more ZK-style. Notably:
   - Local selections are green
   - Buildings are squares, not cornerrects
   - Selection highlighting doesn't float above ground

The main visual difference from unit_shapes is that the selection lines are solid, not gradients. As discussed in Discord, it is plausible to write a fragment shader to replicate this style, but I'm not sure it's worth the effort - I actually like the look of this.

Video in action here: https://clipchamp.com/watch/PjEF4vvYxxJ

moreginger committed:
feat: Copy BAR selected units widget

moreginger committed:
fix: Apply correction to addRadius to get consistent line thickness so far as we can.

For example in the extreme case of a regular triangle we actually need to double addRadius to maintain the same line thickness.

moreginger committed:
feat: Make this look more like ZK selections style.

moreginger committed:
feat: Add preselected units.
These are rendered in the team color but get animation to highlight them as a preselection.

moreginger committed:
feat: Add ally selections and local selection colors

moreginger committed:
feat: Options

moreginger committed:
chore: Tidy up some vars

moreginger committed:
chore: Tested this and it makes no difference.

moreginger committed:
feat: Being strategic about updates is noticeably faster

moreginger committed:
chore: Spring imports and tidyup

moreginger commented on LuaUI/Widgets/Include/instancevbotable.lua:
```diff
@@ -104,17 +104,20 @@ function makeInstanceVBOTable(layout, maxElements, myName, unitIDattribID)
 			if index then 
```
I copied in the current BAR version

moreginger commented on LuaUI/Widgets/Shaders/DrawPrimitiveAtUnit.vert.glsl:
```diff
@@ -0,0 +1,107 @@
+	v_rotationY = atan(modelMatrix[0][2], modelMatrix[0][0]); // we can get the euler Y rot of the model from the model matrix
+	v_uvoffsets = uvoffsets;
+	v_parameters = parameters;
+	// teamID 255 is used for local selection color
```
This is my addition.  I didn't want to use up one of the few parameters for this. BAR uses the default white for selections, I think.

moreginger commented on LuaUI/Widgets/Shaders/DrawPrimitiveAtUnit.vert.glsl:
```diff
@@ -0,0 +1,107 @@
+	v_centerpos = vec4( modelMatrix[3].xyz, 1.0); // We are going to pass the centerpoint to the GS
+	v_lengthwidthcornerheight = lengthwidthcornerheight;
+	#if (ANIMATION == 1)
+	    // No animation when parameters.y is 0
```
Also my addition. I wanted to use the same shader for local selections and ally selections, but animate local selections only.

moreginger committed:
feat: Don't do initial growth animation when paused since it's linked to game frame

moreginger committed:
fix: Don't accidentally hardcode the radius correction to 1 for other polygons.
Also actually limit to 64 vertices. If it were necessary to reduce this to 62 to get 64 then we'd have to do the same for all other polygons.

GoogleFrog commented:
Some comments before I have more time to test:
- Could the ally selections be drawn underneath the self selections? It looks like the blue circles take chunks out of the green ones when it should probably be the other way around.
- From here it looks like the structure selections are too large. They're supposed to match the footprint.

![image](https://github.com/moreginger/Zero-K/assets/1318249/ddff5117-f3c2-4f0a-8931-f6d4d5bd9237)

GoogleFrog commented:
Where would I mess with the circle radius and line thickness?

moreginger commented:
> Where would I mess with the circle radius and line thickness?

Line thickness is one of the options: https://github.com/moreginger/Zero-K/pull/4/files#diff-118899815ddfd865c619a295f7efd8422388b90b2df881bb20970c926485da03R202

The unit sizing parameters including building padding are calculated here: https://github.com/moreginger/Zero-K/pull/4/files#diff-118899815ddfd865c619a295f7efd8422388b90b2df881bb20970c926485da03R58

I haven't touched the latter from the BAR widget.

FWIW I like the slightly larger unit circles as I think it's more aesthetically pleasing when the circles join.

moreginger committed:
feat: Overlay local selections on ally selections so that they are more consistently rendered and visible.

moreginger committed:
feat: Make breathe start from intial size rather than centre on it.
Also make breathe constant size rather than a multiple so that it doesn't misalign selection boxes.

moreginger committed:
feat: Tweak unit sizes to more directly match grid.

moreginger commented on LuaUI/Widgets/gui_selectedunits_gl4.lua:
```diff
@@ -0,0 +1,384 @@
+	timeSinceLastFrame = timeSinceLastFrame + dt
+	SetPausedHack(currentFrame)
+
+	-- TODO: Add a callin for when ally selections change?
```
Do you think this would be useful?

moreginger commented on LuaUI/Widgets/gui_selectedunits_gl4.lua:
```diff
@@ -0,0 +1,384 @@
+	return localSelectionVBO ~= nil and selectionShader ~= nil and otherSelectionVBO ~= nil
+end
+
+local function SetPausedHack(gameFrame)
```
Unfortunate.

moreginger committed:
chore: Reformat

moreginger committed:
fix: Fix hover unit preselection

moreginger committed:
fix: Breathe size should match ally selection at minimum
